source/modulo_components/modulo_components/component_interface.py

```diff
@@ -421,6 +423,61 @@ def add_input(self, signal_name: str, subscription: Union[str, Callable], messag
         except Exception as e:
             self.get_logger().error(f"Failed to add input '{signal_name}': {e}")
 
+    def add_service(self, service_name: str, callback: Union[Callable[[], dict], Callable[[str], dict]]):
+        """
+        Add a service to trigger a callback function.
+        The callback should take either no arguments (empty service) or a single string argument (string service).
+        The string payload can have an arbitrary format to parameterize and control the callback behaviour as desired.
+        It is the responsibility of the service callback to parse the string according to some payload format.
+        When adding a service with a string payload, be sure to document the payload format appropriately.
+
+        :param service_name: The name of the service
+        :param callback: The callback function to execute
+        :return: A dict with the outcome of the service call, containing "success" and "message" fields in the format
+            {"success": [True | False], "message": "..."}
+        """
+        def callback_wrapper(request, response, cb):
+            try:
+                if hasattr(request, "payload"):
+                    ret = cb(request.payload)
+                else:
+                    ret = cb()
+                response.success = True
+
+                # handle different return types
+                if isinstance(ret, dict):
+                    if "success" in ret.keys():
+                        response.success = ret["success"]
+                    else:
+                        # if the dict does not contain a success field, but the callback
+                        # completes without error, assume it was successful
+
+                        response.success = True
+                    if "message" in ret.keys():
+                        response.message = ret["message"]
+                elif isinstance(ret, bool):
+                    response.success = ret
+                else:
+                    response.success = True
```

Comment:
This can be simplified a little

```suggestion

                # if the return does not contain a success field or bool result,
                # but the callback completes without error, assume it was successful
                response.success = True
                if isinstance(ret, dict):
                    if "success" in ret.keys():
                        response.success = ret["success"]
                    if "message" in ret.keys():
                        response.message = ret["message"]
                elif isinstance(ret, bool):
                    response.success = ret
```